### PR TITLE
draw route with speed based color

### PR DESCRIPTION
**Pull request by jcardus**
Draw the route with intepolated color.
Changed MapRoutePoints arrow to text instead of icon so that we can change the color.
I tried to use the palette instead of the hard coded yellow and red colors, but it didn't look good.

Captura de ecrã 2024-06-16, às 20 00 39

### CODE REVIEW

File: src/common/util/colors.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { hexToRgb, rgbToHex } from '@mui/material';
+
+export const interpolateColor = (color1, color2, factor) => {
+  if (factor > 1) factor = 1;
+  if (factor < 0) factor = 0;
+
+  const parseRgb = (rgb) => rgb.match(/\d+/g).map(Number);
```
Comment (tananaev) [inline src/common/util/colors.js:7]: Why do we need this?

File: src/map/MapRoutePath.js [hunk 2]
```diff
@@ -79,17 +81,18 @@ const MapRoutePath = ({ name, positions, coordinates }) => {
...
-        coordinates,
-      },
-      properties: {
-        name,
-        color: reportColor,
-      },
-    });
+    const maxSpeed = positions.map((p) => p.speed).reduce((a, b) => Math.max(a, b), -Infinity);
+    const features = [];
+    for (let i = 0; i < positions.length - 1; i += 1) {
```
Comment (tananaev) [inline src/map/MapRoutePath.js:86]: What happens if there's a single point?

File: src/map/MapRoutePath.js [hunk 3]
```diff
@@ -79,17 +81,18 @@ const MapRoutePath = ({ name, positions, coordinates }) => {
...
+      const p1 = positions[i];
+      const p2 = positions[i + 1];
+      features.push(lineString(
```
Comment (tananaev) [inline src/map/MapRoutePath.js:89]: What does `lineString` function do?

File: src/map/MapRoutePoints.js [hunk 4]
```diff
@@ -27,11 +28,13 @@ const MapRoutePoints = ({ positions, onClick }) => {
...
       source: id,
+      paint: {
+        'text-color': ['get', 'color'],
+      },
       layout: {
-        'icon-image': 'arrow',
-        'icon-allow-overlap': true,
-        'icon-rotate': ['get', 'rotation'],
-        'icon-rotation-alignment': 'map',
+        'text-field': '▲',
```
Comment (tananaev) [inline src/map/MapRoutePoints.js:35]: If we don't use the arrow anymore, we should probably remove it.

File: src/common/util/colors.js (hunk 1 above)
Comment (jcardus) [reply src/common/util/colors.js:7]: hexToRgb returns a string, we can change it to this:

const parseRgb = (rgb) => rgb.split(',').map(Number);

File: src/map/MapRoutePath.js (hunk 2 above)
Comment (jcardus) [reply src/map/MapRoutePath.js:86]: nothings is drawn

File: src/map/MapRoutePath.js (hunk 3 above)
Comment (jcardus) [reply src/map/MapRoutePath.js:89]: it's a helper from turf, creates a lineString feature

File: src/map/MapRoutePoints.js (hunk 4 above)
Comment (jcardus) [reply src/map/MapRoutePoints.js:35]: ok

File: src/common/util/colors.js (hunk 1 above)
Comment (jcardus) [reply src/common/util/colors.js]: I removed it, please check

File: src/map/MapRoutePoints.js (hunk 4 above)
Comment (jcardus) [reply src/map/MapRoutePoints.js:35]: I removed the arrow

File: src/map/MapRoutePath.js (hunk 2 above)
Comment (tananaev) [reply src/map/MapRoutePath.js:86]: But there should be one point, right?

File: src/map/MapRoutePath.js (hunk 3 above)
Comment (tananaev) [reply src/map/MapRoutePath.js:89]: I thought it's a trivial thing to create a line. What does the result look like?

File: src/map/MapRoutePath.js (hunk 2 above)
Comment (jcardus) [reply src/map/MapRoutePath.js:86]: It will be in the other layer (the arrow)

Comment (tananaev) [reply src/map/MapRoutePath.js:86]: I see. That makes sense then.

File: src/map/MapRoutePath.js (hunk 3 above)
Comment (jcardus) [reply src/map/MapRoutePath.js:89]: It is, it's just a helper, it creates the whole structure, including the properties (like it was done before manually):

{
      type: 'Feature',
      geometry: {
        type: 'LineString',
        coordinates: [[....]]
      },
      properties: {
        color: '...',
      },
};

Comment (tananaev) [reply src/map/MapRoutePath.js:89]: I prefer to keep it as JSON instead of using helper.

Comment (jcardus) [reply src/map/MapRoutePath.js]: ok, lineString is gone.

File: src/common/util/colors.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  const c2 = decomposeColor(color2).values;
+
+  const r = Math.round(c1[0] + factor * (c2[0] - c1[0]));
+  const g = Math.round(c1[1] + factor * (c2[1] - c1[1]));
+  const b = Math.round(c1[2] + factor * (c2[2] - c1[2]));
+
+  return `rgb(${r}, ${g}, ${b})`;
+};
+
+export const getSpeedColor = (speed, max) => interpolateColor('#FFFF00', '#FF0000', speed / max);
```
[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js:17]: Before I was receiving the colors in the getSpeedColor

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js:17]: It doesn't answer the question though.

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js:17]: You didn't comment on this before. 
Yes, it answers the question, the getSpeedColor is called inside the components, the theme is available there.

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js:17]: OK, so can we update it then?

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js:17]: I tried a lot of combinations from the theme color, the one it's less bad is warning to success. But the colors are very "near" from each other so you don't get the same effect.

I didn't find any option in the palette to replace the yellow. 

Here's an example:

Screenshot 2024-06-17 at 01 13 36

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js]: Shouldn't the success be green?

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js]: Sorry, I meant warning to error

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js]: What does it look like with success to error?

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js]: Captura de ecrã 2024-06-17, às 04 46 20

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js]: Maybe we can mix 3 colors? Success, warning and error.

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js]: ok, and maybe we can reverse it. In traffic, usually green is moving and red is stoped

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js]: Good point. I guess it depends on what most people use it for.

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js]: error -> warning -> success
Screenshot 2024-06-17 at 18 36 28

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js]: Looks good, but I wonder if it makes sense to reverse colors or not

[the remark this answers is not in the document]
Comment (jcardus) [reply src/common/util/colors.js]: success -> warrning -> error
image

**Comment by ivanfmartinez** [comment]
Very nice to see speed differences, but interpolating with current max speeds can have different colors when looking for different data. 

Maybe this can be a option to use max speed of view or a fixed configurable number per device because 80 km/h can be very good to see in red for a small truck, but not for a standard car that can run at 120 km/h on roads.

Or if still using the max speed at least show it to give to the user a notion about how the colors are distributed.

**Comment by ivanfmartinez** [comment]
personally I like the success -> warrning -> error version 

green on low speed, red on higher speed

File: src/map/MapRoutePath.js [hunk 6]
```diff
@@ -1,7 +1,9 @@
 import { useTheme } from '@mui/styles';
 import { useId, useEffect } from 'react';
 import { useSelector } from 'react-redux';
+import { featureCollection } from '@turf/turf';
```
Comment (tananaev) [inline src/map/MapRoutePath.js:4]: Can we expand `featureCollection` as well? I'm sure it's also a very simple json.

**Comment by tananaev** [comment]
Merged, thank you.

**Comment by Kalabint** [comment]
Here is another suggestion: Turbo Color Ramp.

image

It would look something like this (Range: 0-200km/h, divided into 100 Color Steps):

From Cooler Colors for slower movement to Green in the Middle (about 80km/h) to Yellow/Orange (120km/h) with Red for high speeds.

chrome_WeGyuyZHhT
chrome_J4A4CSvSvD

**Comment by Kalabint** [comment]
The Information density with the MUI Color Scheme is not so great, because the second half is the mapping from Orange to Red, which are very similar colors.

Using the Turbo Color Ramp it would look something like this in Traccar:

image

**Comment by jcardus** [comment]
@Kalabint I agree, but I think what's needed here now it's a legend. @tananaev what do you think?

**Comment by tananaev** [comment]
Don't have a very strong opinion. I'm more worried about quality and reliability of the solution.

**Pull request merged**